File: Backtracking/Python/Letter Combinations of a Phone Number.py
# Problem link: https://leetcode.com/problems/letter-combinations-of-a-phone-number/description/
# Time: O(n * 4 ^ n)
# Space: O(n)
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        store = {'2' : 'abc', '3' : 'def', '4' : 'ghi', 
                '5' : 'jkl', '6' : 'mno', '7' : 'pqrs', 
                '8' : 'tuv', '9' : 'wxyz'}

        def backtrack(i, curr):
            if i == len(digits):
                result.append(curr)
                return
            
            for c in store[digits[i]]:
                backtrack(i + 1, curr + c)
        
        result = []
        if digits:
            backtrack(0, '')
            
        return result

File: Backtracking/Java/Letter Combinations of a Phone Number.java
// Problem link: https://leetcode.com/problems/letter-combinations-of-a-phone-number/description/
// Time: O(n * 4 ^ n)
// Space: O(n)
public class Solution {
    public List<String> letterCombinations(String digits) {
        int n = digits.length();

        String[] store = {"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};

        List<String> result = new ArrayList<>();

        if(n > 0) {
            backtrack(0, digits.toCharArray(), new char[n], result, store);
        }

        return result;
    }

    private void backtrack(int index, char[] digits, char[] curr, List<String> result, String[] store) {
        if(index == digits.length) {
            result.add(new String(curr));
            return;
        }

        int num = digits[index] - '0' - 2;
        for(char c : store[num].toCharArray()) {
            curr[index] = c;
            backtrack(index + 1, digits, curr, result, store);
        }
    }
}
